File: app/src/main/java/co/pushe/sample/as/activities/MainActivity.java
package co.pushe.sample.as.activities;

import android.annotation.SuppressLint;
import android.content.pm.PackageManager;
import androidx.annotation.NonNull;
import androidx.core.util.Consumer;
import androidx.appcompat.app.AppCompatActivity;
import android.os.Bundle;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.appcompat.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ScrollView;
import android.widget.TextView;
import android.widget.Toast;


import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import butterknife.BindView;
import butterknife.ButterKnife;
import co.pushe.plus.Pushe;
import co.pushe.plus.analytics.PusheAnalytics;
import co.pushe.plus.inappmessaging.InAppMessage;
import co.pushe.plus.inappmessaging.PusheInAppMessaging;
import co.pushe.plus.inappmessaging.PusheInAppMessagingListener;
import co.pushe.plus.notification.NotificationButtonData;
import co.pushe.plus.notification.NotificationData;
import co.pushe.plus.notification.PusheNotification;
import co.pushe.plus.notification.PusheNotificationListener;
import co.pushe.plus.notification.UserNotification;
import co.pushe.sample.as.R;
import co.pushe.sample.as.eventbus.MessageEvent;
import co.pushe.sample.as.utils.Stuff;

import static co.pushe.sample.as.utils.Stuff.addText;
import static co.pushe.sample.as.utils.Stuff.alert;
import static co.pushe.sample.as.utils.Stuff.prompt;

/**
 * For further information Go to <a href="https://pushe.co/docs">Docs</a>
 *
 * @author Mahdi Malvandi
 */
@SuppressWarnings({"rawtypes", "unchecked", "unused", "RedundantSuppression"})
@SuppressLint("SetTextI18n,NonConstantResourceId")
public class MainActivity extends AppCompatActivity {

    private PusheInAppMessaging inAppMessaging;

    @BindView(R.id.status)
    TextView status;
    @BindView(R.id.list)
    RecyclerView list;
    @BindView(R.id.toolbar)
    Toolbar toolbar;
    @BindView(R.id.statusContainer)
    ScrollView scroll;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        ButterKnife.bind(this);
        setSupportActionBar(toolbar);
        inAppMessaging = Pushe.getPusheService(PusheInAppMessaging.class);

        try {
            toolbar.setSubtitle("v" + getPackageManager().getPackageInfo(getPackageName(), 0).versionName );
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
        }

        status.setText("Click an action to test it.");

        setupList();

        // Clear on long click
        status.setOnLongClickListener(new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View view) {
                status.setText("Click an action to test it.\n");
                return true;
            }
        });

        status.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                scroll.fullScroll(View.FOCUS_DOWN);
            }
        });

        // Notification listener
        PusheNotification pnotif = Pushe.getPusheService(PusheNotification.class);
        if (pnotif != null) {
            Pushe.getPusheService(PusheNotification.class).setNotificationListener(new PusheNotificationListener() {
                @Override
                public void onNotification(@NonNull NotificationData notification) {
                    addText(scroll, status, "Listener on receiving notification triggered\n" + notification.getTitle());
                }

                @Override
                public void onCustomContentNotification(@NonNull Map<String, Object> customContent) {
                    addText(scroll, status, "Listener on receiving notification with custom content triggered");
                }

                @Override
                public void onNotificationClick(@NonNull NotificationData notification) {
                    addText(scroll, status, "Listener on clicking on notification triggered");
                }

                @Override
                public void onNotificationDismiss(@NonNull NotificationData notification) {
                    addText(scroll, status, "Listener on dismissing notification triggered");
                }

                @Override
                public void onNotificationButtonClick(@NonNull NotificationButtonData button, @NonNull NotificationData notification) {
                    addText(scroll, status, "Listener on clicking on notification button triggered");
                }
            });
        }
        // Piam listener
        PusheInAppMessaging piam = Pushe.getPusheService(PusheInAppMessaging.class);
        if (piam != null) {
            piam.setInAppMessagingListener(new PusheInAppMessagingListener() {
                @Override
                public void onInAppMessageReceived(@NonNull InAppMessage inAppMessage) {
                    addText(scroll, status, "InAppMessage received:\nTitle: " + inAppMessage.getTitle() + "\nContent: " + inAppMessage.getContent());
                }

                @Override
                public void onInAppMessageTriggered(@NonNull InAppMessage inAppMessage) {
                }

                @Override
                public void onInAppMessageClicked(@NonNull InAppMessage inAppMessage) {
                }

                @Override
                public void onInAppMessageDismissed(@NonNull InAppMessage inAppMessage) {
                }

                @Override
                public void onInAppMessageButtonClicked(@NonNull InAppMessage inAppMessage, int piamButtonIndex) {
                }
            });
        }
    }

    /**
     * Set up recyclerView and add items to it.
     * Checkout class {@link Stuff} to see all util methods.
     */
    private void setupList() {
        list.setHasFixedSize(true);
        list.setLayoutManager(new LinearLayoutManager(this));
        list.setAdapter(new Adapter(
                Stuff.listOf(
                        "IDs",
                        "Custom Id",
                        "Phone Number",
                        "Email",
                        "Modules initialization status",
                        "Device registration status",
                        "Topic",
                        "Tag (name:value)",
                        "Analytics: Event",
                        "Analytics: E-commerce",
                        "Notification: AndroidId",
                        "Notification: GoogleAdId",
                        "Notification: CustomId",
                        "Piam: Dismiss All InApps",
                        "Piam: Trigger event"
                ),
                handleItemClick()));
    }

    /**
     * Each clicked item have to do something when clicked.
     * Adapter takes an interface and calls it when an item was clicked.
     * This function does the work and returns an interface.
     */
    private ItemClickListener handleItemClick() {
        return new ItemClickListener() {
            @Override
            public void onItemClick(View v, int position) {
                switch (position) {
                    case 0:
                        alert(MainActivity.this, "IDs", "AndroidId:\n" + Pushe.getDeviceId() + "\nGoogleAdId:\n" + Pushe.getAdvertisingId());
                        break;
                    case 1:
                        prompt(MainActivity.this, "New custom id", "current custom id:" + Pushe.getCustomId(), new Consumer<String>() {
                            @Override
                            public void accept(String s) {
                                Pushe.setCustomId(s);
                                addText(scroll, status, "Custom id is: " + Pushe.getCustomId());
                            }
                        });

                        break;
                    case 2:
                        prompt(MainActivity.this, "New phone number", "current phone number is:" + Pushe.getUserPhoneNumber(), new Consumer<String>() {
                            @Override
                            public void accept(String s) {
                                Pushe.setUserPhoneNumber(s);
                                addText(scroll, status, "phone number is: " + Pushe.getUserPhoneNumber());
                            }
                        });
                        break;
                    case 3:
                        prompt(MainActivity.this, "New Email", "current email is:" + Pushe.getUserEmail(), new Consumer<String>() {
                            @Override
                            public void accept(String s) {
                                Pushe.setUserEmail(s);
                                addText(scroll, status, "user email is: " + Pushe.getUserEmail());
                            }
                        });

                        break;
                    case 4:
                        addText(scroll, status, "modules initialized: " + Pushe.isInitialized());
                        break;
                    case 5:
                        addText(scroll, status, "Device registered: " + Pushe.isRegistered());
                        break;

                    case 6:
                        prompt(MainActivity.this, "Topic", "Topics:[" + Pushe.getSubscribedTopics().toString() + "]\nEnter topic name to subscribe or unsubscribe", "topic name", "Subscribe", "Unsubscribe", new Stuff.Callback<String>() {
                            @Override
                            public void onPositiveButtonClicked(String s) {
                                Pushe.subscribeToTopic(s);
                                addText(scroll, status, "Subscribe to topic: " + s);
                            }

                            @Override
                            public void onNegativeButtonClicked(String s) {
                                Pushe.unsubscribeFromTopic(s);
                                addText(scroll, status, "Unsubscribe from topic: " + s);
                            }
                        });

                        break;

                    case 7:
                        prompt(MainActivity.this, "Tags", "Tags:" + Pushe.getSubscribedTags().toString() + "\nTag in name:value format (add)\n Tag in name1,name2 format (remove)", "topic name", "Add", "Remove", new Stuff.Callback<String>() {
                            @Override
                            public void onPositiveButtonClicked(String s) {
                                if (!s.contains(":")) return;
                                String[] keyValue = s.split(":");
                                if (keyValue.length != 2) return;
                                Map map = new HashMap();
                                map.put(keyValue[0], keyValue[1]);
                                Pushe.addTags(map);
                                addText(scroll, status, "Tag '" + keyValue[0] + "' added ");
                            }

                            @Override
                            public void onNegativeButtonClicked(String s) {
                                String[] keyList = s.split(",");
                                if (keyList.length == 0) return;
                                Pushe.removeTags(Arrays.asList(keyList));
                                addText(scroll, status, "Tags '" + s + "' removed");
                            }
                        });

                        break;

                    case 8:
                        prompt(MainActivity.this, "Event", "Type event name to send", new Consumer<String>() {
                            @Override
                            public void accept(String s) {
                                Pushe.getPusheService(PusheAnalytics.class).sendEvent(s);
                                addText(scroll, status, "Sending event: " + s);

                            }
                        });

                        break;

                    case 9:
                        prompt(MainActivity.this, "E-commerce", "Enter value in name:price format to send data", new Consumer<String>() {
                            @Override
                            public void accept(String s) {
                                if (!s.contains(":")) return;
                                String[] keyValue = s.split(":");
                                if (keyValue.length != 2) return;
                                try {
                                    Pushe.getPusheService(PusheAnalytics.class).sendEcommerceData(keyValue[0], Double.parseDouble(keyValue[1]));
                                    addText(scroll, status, "Sending E-commerce data with name "+keyValue[0]+" and price "+keyValue[1]);
                                } catch (NumberFormatException e) {
                                    addText(scroll, status, "Enter valid price (price should be double)");
                                }
                            }
                        });

                        break;
                    case 10:
                        prompt(MainActivity.this, "Notification", "Enter androidId to send a notification to that user", "androidId",  "Send to ...","Send to me", new Stuff.Callback<String>() {
                            @Override
                            public void onPositiveButtonClicked(String s) {
                                UserNotification userNotification = UserNotification.withAndroidId(s);
                                userNotification.setTitle("title1");
                                userNotification.setContent("content1");
                                Pushe.getPusheService(PusheNotification.class).sendNotificationToUser(userNotification);
                                addText(scroll, status, "Sending notification to AndroidId: " + s);
                            }

                            @Override
                            public void onNegativeButtonClicked(String s) {
                                UserNotification userNotification = UserNotification.withAndroidId(Pushe.getDeviceId());
                                userNotification.setTitle("title1");
                                userNotification.setContent("content1");
                                Pushe.getPusheService(PusheNotification.class).sendNotificationToUser(userNotification);
                                addText(scroll, status, "Sending notification to this device");
                            }
                        });

                        break;

                    case 11:
                        prompt(MainActivity.this, "Notification", "Enter GoogleAdId to send a notification to that user", "GoogleAdId", "Send to ...","Send to me", new Stuff.Callback<String>() {
                            @Override
                            public void onPositiveButtonClicked(String s) {
                                UserNotification userNotification = UserNotification.withAdvertisementId(s);
                                userNotification.setTitle("title1");
                                userNotification.setContent("content1");
                                Pushe.getPusheService(PusheNotification.class).sendNotificationToUser(userNotification);
                                addText(scroll, status, "Sending notification to GoogleAdId: " + s);
                            }

                            @Override
                            public void onNegativeButtonClicked(String s) {
                                UserNotification userNotification = UserNotification.withAdvertisementId(Pushe.getAdvertisingId());
                                userNotification.setTitle("title1");
                                userNotification.setContent("content1");
                                Pushe.getPusheService(PusheNotification.class).sendNotificationToUser(userNotification);
                                addText(scroll, status, "Sending notification to this device");
                            }
                        });

                        break;

                    case 12:
                        prompt(MainActivity.this, "Notification", "Enter custom id to send a notification to that user", "custom id", "Send to ...","Send to me", new Stuff.Callback<String>() {
                            @Override
                            public void onPositiveButtonClicked(String s) {
                                UserNotification userNotification = UserNotification.withCustomId(s);
                                userNotification.setTitle("title1");
                                userNotification.setContent("content1");
                                Pushe.getPusheService(PusheNotification.class).sendNotificationToUser(userNotification);
                                addText(scroll, status, "Sending notification to custom id: " + s);
                            }

                            @Override
                            public void onNegativeButtonClicked(String s) {
                                UserNotification userNotification = UserNotification.withCustomId(Pushe.getCustomId());
                                userNotification.setTitle("title1");
                                userNotification.setContent("content1");
                                Pushe.getPusheService(PusheNotification.class).sendNotificationToUser(userNotification);
                                addText(scroll, status, "Sending notification to this device");
                            }
                        });
                        break;
                    case 13: // InApp Dismiss all
                        if (inAppMessaging == null) return;
                        inAppMessaging.dismissShownInApp();
                        break;
                    case 14: // Trigger an event
                        if (inAppMessaging == null) return;
                        prompt(MainActivity.this, "InAppMessaging", "Enter an event", "", new Consumer<String>() {
                            @Override
                            public void accept(String s) {
                                if (s.isEmpty() || Stuff.listOf("%", "&", "@").contains(s) || s.length() > 10) {
                                    Toast.makeText(MainActivity.this, "Invalid event name", Toast.LENGTH_SHORT).show();
                                } else {
                                    addText(scroll, status, "Triggering event: " + s);
                                    inAppMessaging.triggerEvent(s);
                                }
                            }
                        });
                        break;
                }
            }
        };
    }

    // region EventBus
    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onMessageEvent(MessageEvent event) {
        addText(scroll, status, event.getMessage());
    }

    @Override
    protected void onStart() {
        super.onStart();
        EventBus.getDefault().register(this);
    }

    @Override
    protected void onStop() {
        super.onStop();
        EventBus.getDefault().unregister(this);
    }
    // endregion

    // region List

    // List adapter
    static class Adapter extends RecyclerView.Adapter<Holder> {

        private final List<String> dataSet;
        private final ItemClickListener listener;

        Adapter(List<String> dataSet, ItemClickListener listener) {
            this.dataSet = dataSet;
            this.listener = listener;
        }

        @NonNull
        @Override
        public Holder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
            return new Holder(LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.list_item, viewGroup, false));
        }

        @Override
        public void onBindViewHolder(@NonNull final Holder holder, int i) {
            final int position = i;
            holder.action.setText(dataSet.get(position));
            holder.action.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    listener.onItemClick(view, position);
                }
            });
        }

        @Override
        public int getItemCount() {
            return dataSet.size();
        }
    }

    // List view holder
    static class Holder extends RecyclerView.ViewHolder {

        @BindView(R.id.text)
        TextView action;

        Holder(@NonNull View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }
    }

    interface ItemClickListener {
        void onItemClick(View v, int position);
    }

    // endregion

}

File: app/src/main/java/co/pushe/sample/as/activities/MainActivity.kt
package co.pushe.sample.`as`.activities

import android.annotation.SuppressLint
import android.content.pm.PackageManager
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ScrollView
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.core.util.Consumer
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import butterknife.BindView
import butterknife.ButterKnife
import co.pushe.plus.Pushe
import co.pushe.plus.PusheUser
import co.pushe.plus.analytics.PusheAnalytics
import co.pushe.plus.inappmessaging.InAppMessage
import co.pushe.plus.inappmessaging.PusheInAppMessaging
import co.pushe.plus.inappmessaging.PusheInAppMessagingListener
import co.pushe.plus.notification.*
import co.pushe.sample.`as`.R
import co.pushe.sample.`as`.eventbus.MessageEvent
import co.pushe.sample.`as`.utils.Stuff
import co.pushe.sample.`as`.utils.Stuff.alert
import co.pushe.sample.`as`.utils.Stuff.prompt
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import java.text.DateFormat
import java.util.*

/**
 * For further information Go to [Docs](https://pushe.co/docs)
 *
 * @author Mahdi Malvandi
 */
@Suppress("unused")
@SuppressLint("SetTextI18n,NonConstantResourceId")
class MainActivity : AppCompatActivity() {
    private var inAppMessaging: PusheInAppMessaging? = null

    @BindView(R.id.status)
    lateinit var status: TextView

    @BindView(R.id.list)
    lateinit var list: RecyclerView

    @BindView(R.id.toolbar)
    lateinit var toolbar: Toolbar

    @BindView(R.id.statusContainer)
    lateinit var scroll: ScrollView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        ButterKnife.bind(this)
        setSupportActionBar(toolbar)
        inAppMessaging = Pushe.getPusheService(PusheInAppMessaging::class.java)
        try {
            toolbar.subtitle = "v" + packageManager.getPackageInfo(packageName, 0).versionName
        } catch (e: PackageManager.NameNotFoundException) {
            e.printStackTrace()
        }
        status.text = "Click an action to test it."
        setupList()

        // Clear on long click
        status.setOnLongClickListener {
            status.text = "Click an action to test it.\n"
            true
        }
        status.setOnClickListener { scroll.fullScroll(View.FOCUS_DOWN) }

        // Notification listener
        val pnotif: PusheNotification? = Pushe.getPusheService(PusheNotification::class.java)
        if (pnotif != null) {
            Pushe.getPusheService(PusheNotification::class.java)
                ?.setNotificationListener(object : PusheNotificationListener {
                    override fun onNotification(notification: NotificationData) {
                        addText(
                            """
     Listener on receiving notification triggered
     ${notification.title}
     """.trimIndent()
                        )
                    }

                    override fun onCustomContentNotification(customContent: Map<String, Any>) {
                        addText("Listener on receiving notification with custom content triggered")
                    }

                    override fun onNotificationClick(notification: NotificationData) {
                        addText("Listener on clicking on notification triggered")
                    }

                    override fun onNotificationDismiss(notification: NotificationData) {
                        addText("Listener on dismissing notification triggered")
                    }

                    override fun onNotificationButtonClick(
                        button: NotificationButtonData,
                        notification: NotificationData
                    ) {
                        addText("Listener on clicking on notification button triggered")
                    }
                })
        }
        // Piam listener
        Pushe.getPusheService(PusheInAppMessaging::class.java)
            ?.setInAppMessagingListener(object : PusheInAppMessagingListener {
                override fun onInAppMessageReceived(inAppMessage: InAppMessage) {
                    addText(
                        """
           InAppMessage received:
           Title: ${inAppMessage.title}
           Content: ${inAppMessage.content}
           """.trimIndent()
                    )
                }

                override fun onInAppMessageTriggered(inAppMessage: InAppMessage) {}
                override fun onInAppMessageClicked(inAppMessage: InAppMessage) {}
                override fun onInAppMessageDismissed(inAppMessage: InAppMessage) {}
                override fun onInAppMessageButtonClicked(inAppMessage: InAppMessage, piamButtonIndex: Int) {}
            })
    }

    /**
     * Set up recyclerView and add items to it.
     * Checkout class [Stuff] to see all util methods.
     */
    private fun setupList() {
        list.setHasFixedSize(true)
        list.layoutManager = LinearLayoutManager(this)
        list.adapter = Adapter(menus())
    }

    private fun menus(): Map<String, () -> Unit> {
        return mapOf(
            "IDs" to {
                alert(
                    this@MainActivity, "IDs", """
     AndroidId:
     ${Pushe.getDeviceId()}
     GoogleAdId:
     ${Pushe.getAdvertisingId()}
     """.trimIndent()
                )
            },
            "Custom Id" to {
                prompt(this@MainActivity, "New custom id", "current custom id:" + Pushe.getCustomId(), Consumer { s ->
                    Pushe.setCustomId(s)
                    addText("Custom id is: " + Pushe.getCustomId())
                })
            },
            "Phone Number" to {
                prompt(
                    this@MainActivity,
                    "New phone number",
                    "current phone number is:" + Pushe.getUserPhoneNumber(),
                    Consumer { s ->
                        Pushe.setUserPhoneNumber(s)
                        addText("phone number is: " + Pushe.getUserPhoneNumber())
                    })
            },
            "Email" to {
                prompt(this@MainActivity, "New Email", "current email is:" + Pushe.getUserEmail(), Consumer { s ->
                    Pushe.setUserEmail(s)
                    addText("user email is: " + Pushe.getUserEmail())
                })
            },
            "Modules initialization status" to {
                addText("modules initialized: " + Pushe.isInitialized())
            },
            "Device registration status" to {
                addText("Device registered: " + Pushe.isRegistered())
            },
            "Topic" to {
                prompt(this@MainActivity, "Topic", """
     Topics:[${Pushe.getSubscribedTopics()}]
     Enter topic name to subscribe or unsubscribe
     """.trimIndent(), "topic name", "Subscribe", "Unsubscribe", object : Stuff.Callback<String> {
                    override fun onPositiveButtonClicked(t: String) {
                        Pushe.subscribeToTopic(t)
                        addText("Subscribe to topic: $t")
                    }

                    override fun onNegativeButtonClicked(t: String) {
                        Pushe.unsubscribeFromTopic(t)
                        addText("Unsubscribe from topic: $t")
                    }
                })
            },
            "Tag (name:value)" to {
                prompt(this@MainActivity, "Tags", """Tags:${Pushe.getSubscribedTags()}
Tag in name:value format (add)
 Tag in name1,name2 format (remove)""", "topic name", "Add", "Remove", object : Stuff.Callback<String> {
                    override fun onPositiveButtonClicked(t: String) {
                        if (!t.contains(":")) return
                        val keyValue = t.split(":").toTypedArray()
                        if (keyValue.size != 2) return
                        val map: MutableMap<String, String> = mutableMapOf()
                        map[keyValue[0]] = keyValue[1]
                        Pushe.addTags(map)
                        addText("Tag '" + keyValue[0] + "' added ")
                    }

                    override fun onNegativeButtonClicked(t: String) {
                        val keyList = t.split(",").toTypedArray()
                        if (keyList.isEmpty()) return
                        Pushe.removeTags(listOf(*keyList))
                        addText("Tags '$t' removed")
                    }
                })
            },
            "User: Report login" to {
                Pushe.userLoggedIn("log_data")
                addText("Calling user logIn")
            },
            "User: Report logout" to {
                Pushe.userLoggedOut()
                addText("Calling user logOut")
            },
            "User: Is loggedIn" to {
                addText("User is ${if (Pushe.isUserLoggedIn()) "" else "not "}logged in")
            },
            "User: Sync sample user attribute" to {
                val userToSync = Pushe.getCurrentUserAttributes() ?: PusheUser() // use this way to update current instead of creating a new one
                    .setBirth(12, 2, 1990)
                    .setCity("Sabzevar")
                    .setEmail("hi@pushe.co")
                    .setPostalCode("123445566")
                    .setLocality("tabas-square")
                    .setCompany("Pushe")
                    .setEmailOptedIn(true)
                    .setFirstName("Johnny")
                    .setLastName("Cage")
                Pushe.syncUserAttributes(userToSync)
                addText("Syncing user\n ${userToSync.toMap()}")
            },
            "User: Get current synced attribute" to {
                val user = Pushe.getCurrentUserAttributes()
                addText(user?.toMap()?.toString() ?: "No synced user is saved")
            },
            "Analytics: Event" to {
                prompt(this@MainActivity, "Event", "Type event name to send", Consumer { s ->
                    Pushe.getPusheService(PusheAnalytics::class.java)?.sendEvent(s)
                    addText("Sending event: $s")
                })
            },
            "Analytics: E-commerce" to {
                prompt(this@MainActivity, "E-commerce", "Enter value in name:price format to send data", Consumer { s ->
                    if (!s.contains(":")) return@Consumer
                    val keyValue = s.split(":").toTypedArray()
                    if (keyValue.size != 2) return@Consumer
                    try {
                        Pushe.getPusheService(PusheAnalytics::class.java)
                            ?.sendEcommerceData(keyValue[0], keyValue[1].toDouble())
                        addText(
                            "Sending E-commerce data with name " + keyValue[0] + " and price " + keyValue[1]
                        )
                    } catch (e: NumberFormatException) {
                        addText("Enter valid price (price should be double)")
                    }
                })
            },
            "Notification: AndroidId" to {
                prompt(
                    this@MainActivity,
                    "Notification",
                    "Enter androidId to send a notification to that user",
                    "androidId",
                    "Send to ...",
                    "Send to me",
                    object : Stuff.Callback<String> {
                        override fun onPositiveButtonClicked(t: String) {
                            val userNotification: UserNotification = UserNotification.withDeviceId(t)
                            userNotification.setTitle("title1")
                            userNotification.setContent("content1")
                            Pushe.getPusheService(PusheNotification::class.java)
                                ?.sendNotificationToUser(userNotification)
                            addText("Sending notification to AndroidId: $t")
                        }

                        override fun onNegativeButtonClicked(t: String) {
                            val userNotification: UserNotification = UserNotification.withDeviceId(Pushe.getDeviceId())
                            userNotification.setTitle("title1")
                            userNotification.setContent("content1")
                            Pushe.getPusheService(PusheNotification::class.java)
                                ?.sendNotificationToUser(userNotification)
                            addText("Sending notification to this device")
                        }
                    })
            },
            "Notification: GoogleAdId" to {
                prompt(
                    this@MainActivity,
                    "Notification",
                    "Enter GoogleAdId to send a notification to that user",
                    "GoogleAdId",
                    "Send to ...",
                    "Send to me",
                    object : Stuff.Callback<String> {
                        override fun onPositiveButtonClicked(t: String) {
                            val userNotification: UserNotification = UserNotification.withAdvertisementId(t)
                            userNotification.setTitle("title1")
                            userNotification.setContent("content1")
                            Pushe.getPusheService(PusheNotification::class.java)
                                ?.sendNotificationToUser(userNotification)
                            addText("Sending notification to GoogleAdId: $t")
                        }

                        override fun onNegativeButtonClicked(t: String) {
                            val userNotification: UserNotification =
                                UserNotification.withAdvertisementId(Pushe.getAdvertisingId())
                            userNotification.setTitle("title1")
                            userNotification.setContent("content1")
                            Pushe.getPusheService(PusheNotification::class.java)
                                ?.sendNotificationToUser(userNotification)
                            addText("Sending notification to this device")
                        }
                    })
            },
            "Notification: CustomId" to {
                prompt(
                    this@MainActivity,
                    "Notification",
                    "Enter custom id to send a notification to that user",
                    "custom id",
                    "Send to ...",
                    "Send to me",
                    object : Stuff.Callback<String> {
                        override fun onPositiveButtonClicked(t: String) {
                            val userNotification: UserNotification = UserNotification.withCustomId(t)
                            userNotification.setTitle("title1")
                            userNotification.setContent("content1")
                            Pushe.getPusheService(PusheNotification::class.java)
                                ?.sendNotificationToUser(userNotification)
                            addText("Sending notification to custom id: $t")
                        }

                        override fun onNegativeButtonClicked(t: String) {
                            val userNotification: UserNotification = UserNotification.withCustomId(Pushe.getCustomId())
                            userNotification.setTitle("title1")
                            userNotification.setContent("content1")
                            Pushe.getPusheService(PusheNotification::class.java)
                                ?.sendNotificationToUser(userNotification)
                            addText("Sending notification to this device")
                        }
                    })
            },
            "Piam: Dismiss All InApps" to {
                inAppMessaging?.dismissShownInApp()
            },
            "Piam: Trigger event" to {

                prompt(this@MainActivity, "InAppMessaging", "Enter an event", "", Consumer { s ->
                    if (s.isEmpty() || listOf("%", "&", "@").contains(s) || s.length > 10) {
                        Toast.makeText(this@MainActivity, "Invalid event name", Toast.LENGTH_SHORT).show()
                    } else {
                        addText("Triggering event: $s")
                        inAppMessaging!!.triggerEvent(s)
                    }
                })
            }
        )
    }

    /**
     * Add text instead of replacing the text in android studio.
     * Takes a textView and a text and adds the text to the textView.
     */
    fun addText(text: String) {
        val textView = status
        val scrollView = scroll
        val currentText = textView.text as String
        val currentDateTimeString = DateFormat.getDateTimeInstance().format(Date())
        val newText = "$currentText\n-----\n$text\nTime: $currentDateTimeString\n"
        textView.text = newText
        scrollView.fullScroll(View.FOCUS_DOWN)
    }

    // region EventBus
    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onMessageEvent(event: MessageEvent) {
        addText(event.message)
    }

    override fun onStart() {
        super.onStart()
        EventBus.getDefault().register(this)
    }

    override fun onStop() {
        super.onStop()
        EventBus.getDefault().unregister(this)
    }

    // endregion
    // region List
    // List adapter
    internal class Adapter(private val dataSet: Map<String, () -> Unit>) : RecyclerView.Adapter<Holder?>() {

        override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): Holder {
            return Holder(LayoutInflater.from(viewGroup.context).inflate(R.layout.list_item, viewGroup, false))
        }

        override fun onBindViewHolder(holder: Holder, i: Int) {
            val key = dataSet.keys.elementAt(i)
            holder.action.text = key
            holder.action.setOnClickListener { view -> dataSet[key]?.invoke() }
        }

        override fun getItemCount(): Int {
            return dataSet.size
        }
    }

    // List view holder
    internal class Holder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        @BindView(R.id.text)
        lateinit var action: TextView

        init {
            ButterKnife.bind(this, itemView)
        }
    }

    // endregion
}